app/controllers/users_controller.rb

```diff
@@ -23,4 +56,27 @@ def user_params
     params.require(:user).permit :name, :email, :password,
       :password_confirmation
   end
+
+  def logged_in_user
+    return if logged_in?
+    store_location
+    flash[:danger] = t ".notice"
+    redirect_to login_url
+  end
+
+  def correct_user
+    @user = User.find_by(id: params[:id])
+    return if current_user?(@user)
+    flash[:danger] = t ".danger"
+    redirect_to(root_url)
+  end
+
+  def admin_user
+    redirect_to(root_url) unless current_user.admin?
+  end
+
+  def find_user
+    @user = User.find_by id: params[:id]
+    redirect_to root_path if @user.nil?
```

Comment:
em đổi lại xử lý logic như này nhé:
```suggestion
    redirect_to root_path unless @user
```